### ISSUE TITLE
Refactor Mixtral to reuse code from MegaBlocks

### ISSUE THREAD

**Issue by tgale96** [vllm-project/vllm#2032]
Hello! A [portion](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/mixtral.py#L223-L326) of the MoE implementation for Mixtral is copied directly from MegaBlocks. It's somewhat error prone code and I've been meaning to factor out helpers for it, which we could reuse to avoid having this duplicated in vLLM. If this is interesting to you I'll send a PR :)

**Comment by WoosukKwon** [comment]
Hi @tgale96, sounds wonderful! We'd appreciate your contribution!

**Comment by tgale96** [comment]
Great :) I'll get to it this week, I think!

No, we won't recompile when the batch size changes. We will only re-compile when the [constexpr parameters](https://github.com/stanford-futuredata/stk/blob/main/stk/backend/triton_kernels.py) of the kernels change, which are implementation details like the tile dimensions and threadblock-swizzle config.

**Comment by WoosukKwon** [comment]
Hi @tgale96, we've just merged #2090, which doesn't use Megablocks but seems to solve the performance (and ROCm compatibility) issue at the moment. However, we'd love to see your PR since the current method might not be efficient for other future MoE models.

**Comment by tgale96** [comment]
Hi! The PR is pretty simple in vLLM - those functions I highlighted we'll exposed as public APIs, we'll delete them from vLLM and then we'll call into megablocks for them. If you'd like to see it I'm happy to make the changes, but if vLLM won't use the APIs we will probably avoid exposing them so we are free to change them in the future :)

I'm very curious to understand that PR better! I dropped a question over there :)
